### PR TITLE
[GLUTEN-5696] Add preprojection support for ArrowEvalPythonExec

### PR DESCRIPTION
**Pull request by yma11**
## What changes were proposed in this pull request?

Add preprojection support for ArrowEvalPythonExec
Fixes: \#5696)

## How was this patch tested?

New UTs added

### CODE REVIEW

**Comment by yma11** [comment]
@jinchengchenghh @zhouyuan please help take a review. Thanks.

File: gluten-core/src/main/scala/org/apache/gluten/backendsapi/SparkPlanExecApi.scala [hunk 1]
```diff
@@ -738,5 +739,7 @@ trait SparkPlanExecApi {
 
   def genPostProjectForGenerate(generate: GenerateExec): SparkPlan
 
+  def genPreProjectForArrowEvalPythonExec(arrowEvalPythonExec: ArrowEvalPythonExec): SparkPlan
```
[the remark this answers is not in the document]
Comment (yma11) [reply gluten-core/src/main/scala/org/apache/gluten/backendsapi/SparkPlanExecApi.scala]: Updated.

File: backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala [hunk 2]
```diff
@@ -279,16 +282,30 @@ case class ColumnarArrowEvalPythonExec(
...
         val dataTypes = new ArrayBuffer[DataType]
+        val originalOffsets = new ArrayBuffer[Int]
         val argOffsets = inputs.map {
           input =>
             input.map {
               e =>
                 if (allInputs.exists(_.semanticEquals(e))) {
                   allInputs.indexWhere(_.semanticEquals(e))
                 } else {
+                  if (!e.isInstanceOf[AttributeReference]) {
```
Comment (jinchengchenghh) [inline backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala:297]: else if (!e.isInstanceOf[AttributeReference]) {}
else {}

File: backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala [hunk 3]
```diff
@@ -299,15 +316,20 @@ case class ColumnarArrowEvalPythonExec(
...
         val inputBatchIter = contextAwareIterator.map {
           inputCb =>
             ColumnarBatches.ensureLoaded(ArrowBufferAllocators.contextInstance, inputCb)
-            // 0. cache input for later merge
             ColumnarBatches.retain(inputCb)
+            // 0. cache input for later merge
             inputCbCache += inputCb
-            inputCb
+            var colsForEval = new ArrayBuffer[ColumnVector]()
+            for (i <- originalOffsets) {
```
Comment (jinchengchenghh) [inline backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala:329]: Why generate a new cb here? Can you add a comment here?

File: backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala [hunk 4]
```diff
@@ -335,6 +357,65 @@ case class ColumnarArrowEvalPythonExec(
...
+  }
+
+  def pullOutPreProject(arrowEvalPythonExec: ArrowEvalPythonExec): SparkPlan = {
+    // pull out preproject
+    val (_, inputs) = arrowEvalPythonExec.udfs.map(collectFunctions).unzip
+    val expressionMap = new mutable.HashMap[Expression, NamedExpression]()
+    // flatten all the arguments
+    val allInputs = new ArrayBuffer[Expression]
+    for (input <- inputs) {
+      input.map {
```
Comment (jinchengchenghh) [inline backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala:402]: input.filter(allInputs.exists(_.semanticEquals(e))

File: backends-velox/src/test/scala/org/apache/gluten/execution/python/ArrowEvalPythonExecSuite.scala [hunk 5]
```diff
@@ -58,4 +58,47 @@ class ArrowEvalPythonExecSuite extends WholeStageTransformerSuite {
...
+      ("2", 1, "2", 2),
+      ("2", 2, "2", 4),
+      ("3", 1, "3", 2),
+      ("3", 2, "3", 4),
+      ("0", 1, "0", 2),
+      ("3", 0, "3", 0)
+    ).toDF("a", "b", "p_a", "d_b")
+
+    val df = base.withColumn("p_a", pyarrowTestUDF(base("a"))).withColumn("d_b", base("b") * 2)
+    df.explain()
```
Comment (jinchengchenghh) [inline backends-velox/src/test/scala/org/apache/gluten/execution/python/ArrowEvalPythonExecSuite.scala:78]: remove

File: backends-velox/src/test/scala/org/apache/gluten/execution/python/ArrowEvalPythonExecSuite.scala [hunk 6]
```diff
@@ -58,4 +58,47 @@ class ArrowEvalPythonExecSuite extends WholeStageTransformerSuite {
...
+      ("3", 1, 2, "3", 2),
+      ("3", 2, 4, "3", 4),
+      ("0", 1, 2, "0", 2),
+      ("3", 0, 0, "3", 0)
+    ).toDF("a", "b", "d_b", "p_a", "p_b")
+    val df = base
+      .withColumn("d_b", base("b") * 2)
+      .withColumn("p_a", pyarrowTestUDF(base("a")))
+      .withColumn("p_b", pyarrowTestUDF(base("b") * 2))
+    df.explain()
```
Comment (jinchengchenghh) [inline backends-velox/src/test/scala/org/apache/gluten/execution/python/ArrowEvalPythonExecSuite.scala:101]: remove

File: backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala (hunk 4 above)
Comment (yma11) [reply backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala:402]: This line is not to filter the inputs but add the matched expression in `allInputs` and replace it using faked `Attribute`.

**Review by jinchengchenghh — APPROVED** [review]

**Review by zhztheplayer — APPROVED** [review]
Thanks!

**Pull request merged**